modules/cache/jvm/src/main/scala/coursier/cache/CacheUrl.scala

```diff
@@ -20,21 +20,34 @@ object CacheUrl {
 
   private val handlerClsCache = new ConcurrentHashMap[String, Option[URLStreamHandler]]
 
-  private def handlerFor(url: String): Option[URLStreamHandler] = {
+  private def handlerFor(url: String, classLoaders: Seq[ClassLoader]): Option[URLStreamHandler] = {
     val protocol = url.takeWhile(_ != ':')
 
     Option(handlerClsCache.get(protocol)) match {
       case None =>
-        val clsName = s"coursier.cache.protocol.${protocol.capitalize}Handler"
+        val clsName = List(
+          "coursier",
+          "cache",
+          "protocol",
+          s"${protocol.capitalize}Handler",
+        ).mkString(".")
+
         def clsOpt(loader: ClassLoader): Option[Class[_]] =
           try Some(Class.forName(clsName, false, loader))
           catch {
             case _: ClassNotFoundException =>
               None
           }
 
-        val clsOpt0: Option[Class[_]] = clsOpt(Thread.currentThread().getContextClassLoader)
-          .orElse(clsOpt(getClass.getClassLoader))
+        val clsOpt0: Option[Class[_]] = {
+          def fromProvided = classLoaders.toStream.map(clsOpt)
+          def fromThread = clsOpt(Thread.currentThread().getContextClassLoader)
+          def fromClass = clsOpt(getClass.getClassLoader)
+
+          fromProvided.append(List(fromThread, fromClass)).collectFirst {
+            case Some(cls) => cls
+          }
+        }
```

Comment:
I'm not sure this lazily evaluates `fromClass`... Maybe we can use iterators instead?
```suggestion
        val clsOpt0: Option[Class[_]] =
          (classLoaders ++ Iterator(Thread.currentThread().getContextClassLoader, getClass.getClassLoader))
            .flatMap(clsOpt(_).iterator)
            .toStream
            .headOption
```